x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -652,149 +621,209 @@ private GetAliasesRequest createGetAliasesRequest(FieldCapabilitiesResponse resp
-     *           "aggregatable": false,
-     *           "indices": [
-     *               "bar",
-     *               "foo"
-     *           ],
-     *           "non_searchable_indices": [
-     *               "foo"
-     *           ]
-     *       },
-     *       "keyword": {
-     *           "type": "keyword",
-     *           "searchable": false,
-     *           "aggregatable": true,
-     *           "non_aggregatable_indices": [
-     *               "bar", "baz"
-     *           ]
-     *       }
-     *   }
-     */
-    private static Map<String, InvalidMappedField> getInvalidFieldsForAliases(String fieldName, Map<String, FieldCapabilities> types,
-            ImmutableOpenMap<String, List<AliasMetadata>> aliases) {
-        if (aliases == null || aliases.isEmpty()) {
-            return emptyMap();
-        }
+    private static Map<String, InvalidMappedField> validateField(String fieldName, Map<String, QlFieldCapabilities> qlFieldCaps,
+        Map<String, FieldCapabilities> esTypes)
+    {
         Map<String, InvalidMappedField> invalidFields = new HashMap<>();
-        Map<String, Set<String>> typesErrors = new HashMap<>(); // map holding aliases and a list of unique field types across its indices
-        Map<String, Set<String>> aliasToIndices = new HashMap<>(); // map with aliases and their list of indices
 
-        Iterator<ObjectObjectCursor<String, List<AliasMetadata>>> iter = aliases.iterator();
-        while (iter.hasNext()) {
-            ObjectObjectCursor<String, List<AliasMetadata>> index = iter.next();
-            for (AliasMetadata aliasMetadata : index.value) {
-                String aliasName = aliasMetadata.alias();
-                aliasToIndices.putIfAbsent(aliasName, new HashSet<>());
-                aliasToIndices.get(aliasName).add(index.key);
+        // for each indices group
+        for (Entry<String, QlFieldCapabilities> capabilities : qlFieldCaps.entrySet()) {
+            String indicesGroupName = capabilities.getKey();
+            Map<String, QlFieldCapability> types = capabilities.getValue().getFields().get(fieldName);
+            
+            if (types == null || fieldName.startsWith("_")) {
+                continue;
+            }
+            InvalidMappedField invalidField = validate(fieldName, types);
+            if (invalidField != null) {
+                invalidFields.put(indicesGroupName, invalidField);
+            } else if (qlFieldCaps.size() == 1 && qlFieldCaps.containsKey("*") && esTypes.containsKey(KEYWORD.esType())
+                && esTypes.containsKey(CONSTANT_KEYWORD.esType())) {
+                esTypes.remove(CONSTANT_KEYWORD.esType());
             }
         }
 
-        // iterate over each type
-        for (Entry<String, FieldCapabilities> type : types.entrySet()) {
-            String esFieldType = type.getKey();
-            if (esFieldType == UNMAPPED) {
-                continue;
+        if (invalidFields.isEmpty() == false) {
```

Comment:
return `invalidFields.isEmpty() ? emptyMap() : invalidFields` which is a minor optimization since `invalidFields` could be returned always (if it's empty, it has the same behavior as emptyMap)